src/pgl/docs/pgld.1
.\" Modified Thu Jan  7 19:07:17 CET 2010: jre <jre-phoenix@users.sourceforge.net>
.\" Modified several times by jre
.\" Fri Aug 28 16:10:02 GMT 2009: sloter <laurent at sloter.org>
.\"
.\"   This file is part of pgl.
.\"
.\"   pgl is free software: you can redistribute it and/or modify
.\"   it under the terms of the GNU General Public License as published by
.\"   the Free Software Foundation, either version 3 of the License, or
.\"   (at your option) any later version.
.\"
.\"   pgl is distributed in the hope that it will be useful,
.\"   but WITHOUT ANY WARRANTY; without even the implied warranty of
.\"   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\"   GNU General Public License for more details.
.\"
.\"   You should have received a copy of the GNU General Public License
.\"   along with pgl.  If not, see <http://www.gnu.org/licenses/>.
.TH PGL 1 "2010-01-07" "Version 2.0" "pgl Manual"

.SH NAME
\fBpgl\fR \- an IP blocker for Linux.

.SH SYNOPSIS
.B pgld
[\fI\-s\fR] [\fI\-l LOGFILE\fR] [\fI\-d\fR] [\fI\-c CHARSET\fR] [\fI\-p
PIDFILE\fR] [\fI\-q 0-65535\fR] \fI\-r MARK\fR \fI\-a MARK\fR BLOCKLIST(S)
.PP
.B pgld
[\fI\-c CHARSET\fR] \-m [\fIBLOCKLIST(S)\fR]
.PP
.B pglcmd
OPTION
.PP
.B pglcmd
search PATTERN
.PP
.B pglgui

.SH DESCRIPTION
.B pgld
checks internet traffic based on large lists of IP address ranges in order to
protect your privacy.
.PP
.B pglcmd
is designed to do all tasks related to pgl:
.IP \(em 3
Start and stop pgl. Or let either SysV init or systemd do this for you.
.IP \(em 3
Update your blocklists from online sources. Or let cron do this automatically.
Backups will be used if a problem occurs. Additionally you may use your local
blocklists.
.IP \(em 3
Examine your selected blocklists by searching the single blocklists for
keywords.
.IP \(em 3
Remove lines by keyword from the blocklists.
.IP \(em 3
Handle your iptables rules: use a default setup, easily allow all traffic on
specific ports and use an allow list, or add your own sophisticated iptables
rules.
.IP \(em 3
Allow all LAN traffic and the DNS server automatically. If you are on a public
LAN, you probably want to disable this feature.
.IP \(em 3
Check the status and test pgl.
.IP \(em 3
A watchdog monitors pgl.
.IP \(em 3
Detects if kernel modules are needed and loads them if necessary.
.IP \(em 3
Set verbosity and logging options.
.IP \(em 3
Daily rotation of the logfiles.

.SH OPTIONS pgld
.TP
.B \-d
Enable D-Bus support.
.TP
.B \-s
Enable logging to syslog and/or systemd journal.
.TP
.B \-l LOGFILE
Enable logging to LOGFILE.
.TP
.B \-c CHARSET
Specify blocklist file CHARSET.
.TP
.B \-p PIDFILE
Use a PIDFILE.
.TP
.B \-a MARK
Place a 32-bit MARK on accepted packets (packet in not in blocklist).
.TP
.B \-r MARK
Place a 32-bit MARK on rejected packets (packet in in blocklist).
.TP
.B \-q 0-65535
Specify a 16-bit NFQUEUE number. Must match \-\-queue-num used in iptables
rules.
.TP
.B \-m [BLOCKLIST(S)]
Load, sort, merge, and dump blocklist(s) specified or piped from stdin.
.TP
.B BLOCKLIST
BLOCKLIST may be in p2p, p2b, ipfilter.dat or lowmem format. pgld has builtin
support for gz packed lists, if it was compiled with zlib support. pglcmd
unpacks gz, zip and 7z lists for pgld.
.PP

.SH OPTIONS pglcmd
.TP
.B start
inserts iptables rules, starts pgld and the watchdog. If the blocklist
configuration changed, rebuilds the master blocklist.
.TP
.B stop
deletes iptables rules and stops pgld.
.TP
.B start_wd
starts the watchdog. Don´t start pglcmd.wd directly. Remember that per
default the watchdog gets started automatically on start. The watchdog checks in
regular intervals (per default 5 minutes) if some of the necessary iptables
rules exist and if the daemon is responsive. If one of these tests fails, it
restarts pgl. This is especially useful, if you have other firewall applications
running, because they often mess up the iptables rules, which can be fixed by
restarting pgl. Unfortunately this manual action is often forgotten.
.TP
.B stop_wd
stops the watchdog.
.TP
.B restart
restarts pgl.
.TP
.B reload
rebuilds the master blocklist if necessary and reloads pgld if it is running.
.TP
.B update
updates the blocklists, rebuilds the master blocklist and reloads pgld if it is
running.
.TP
.B force-restart
like restart, but forces to rebuild blocklist
.TP
.B force-reload
like reload, but forces to rebuild blocklist
.TP
.B force-update
like update, but forces to download blocklists again
.TP
.B status
gives the iptables settings and the status of pgld and the watchdog.
.TP
.B test
does a simple test to check if pgl is working (pings a random IP from the
blocklist, checks if this IP was logged in the logfile, and if the IP answered).
.TP
.B search
Outputs the occurences of a keyword and the names of the single blocklists.
.TP
.B stats
reports MoBlock's statistics.
.TP
.B reset_stats
resets MoBlock's statistics.
.TP
.B show_config
shows the current configuration settings.
.PP
Note for blocklist operations: When the master blocklist is built, missing
single blocklists are downloaded. If any blocklist fails to download, and if
there is no old version available, the operation aborts. If a downloaded
blocklist fails to extract, it is deleted and the operation aborts.

.SH SIGNALS pgld
.TP
.B USR1
Dump and reset stats.
.TP
.B USR2
Just dump stats.
.TP
.B HUP
Reopen the logfile and reload the blocklist.
.TP
.B TERM
.TP
.B INT
.PP

.SH CONFIGURATION pglcmd
.P
Remote blocklists are configured in \fIBLOCKLISTS_LIST\fR
(/etc/pgl/blocklists.list). Local blocklists are placed in
\fILOCAL_BLOCKLIST_DIR\fR (/etc/pgl/blocklists.local/).
.P
The allow list is \fIallow.p2p\fR (/etc/pgl/allow.p2p). Per default,
the allowlist is used for incoming and outgoing connections. If desired
different allow lists for incoming, outgoing and forward connections may be
used.
.P
The rest is done in \fIpglcmd.conf\fR
(/etc/pgl/pglcmd.conf). Refer to pglcmd.defaults
(/usr/lib/pgl/pglcmd.defaults) for the complete set of possible
configuration variables with comments.

.SH RETURN VALUES pglcmd
\fIpglcmd status\fR gives you the following return values:
.IP
.B 0
pgld is running
.br
.B 1
pgld is dead and pidfile exists
.br
.B 3
pgld is not running
.br
.B 4
pgld is unknown
.PP
\fIpglcmd test\fR gives you the following return values:
.IP
.B 0
test succeeded
.br
.B 1
test failed
.br
.B 2
not testable
.PP

.SH EXIT STATUS pglcmd
.IP
.B 2
pgl was started with bad arguments
.br
.B 4
User is not root
.br
.B 5
pgld is not installed
.br
.B 6
pglcmd is not configured correctly
.br
.B 7
File is missing/has wrong mode
.br
.B 8
iptables command failed
.br
.B 9
blocklist is not available or blocklist archive is corrupted
.br
.B 66
Directory is missing
.br
.B 170
External binary or function is missing
.br
.B 171
Testhost (serving the default blocklists) is not reachable

.SH NOTES
.PP
By default pgl will be started at every system boot up and the blocklists will
be updated once a day.

.SH TECHNICAL NOTE
.PP
pgld checks traffic (packets) that is sent to the iptables NFQUEUE target. If
the necessary netfilter support is not built in the kernel directly, pglcmd will
load the necessary kernel modules. If they are not available, pgld can not be
started.
.PP
If a packet matches the blocklist, then pgld MARKs the packet. Blocked packets
get the MARK "10", which is shown as "0xa" by iptables.
.PP
If a packet doesn't match the blocklist, then pgld MARKs the packet. Allowed
packets get the MARK "20", which is shown as "0x14" by iptables.
.PP
A MARKed packet repeats the hook function (NF_REPEAT). So it is sent back to
the head of the iptables chain again. This means it enters the chain INPUT,
OUTPUT or FORWARD again, but this time bearing a MARK. Then further iptables
rules that match the MARK decide what happens with the packets.
.PP
Per default pglcmd sets iptables rules to REJECT outgoing packets, and to DROP
incoming and forwarded packets, if they were "marked block". If they were
"marked allow", pglcmd's iptables setup just ignores them, so other iptables
rules decide what happens to them.
.PP
It is strongly recommended to use the MARKing feature, because this allows one
to integrate pgl with other firewalls.
.PP
A packet may only bear one MARK, so there mustn't be any other applications or
iptables rules that MARK packets. Otherwise the setup will not work and the
packet will loop forever.
.PP

.SH WARNING: Users with firewall (iptables rules)
\fBpgl\fR does not conflict with other firewalls (iptables rules) as long as you
make sure the following two conditions hold:
.IP \(em 3
Other firewalls must not MARK packets.
.IP \(em 3
pgl is started after other firewalls. If other firewalls are started or reloaded
after pgl, then you may need to restart pglcmd again. You will be fine, if the
iptables rules which send traffic to the iptables chains (pgl_in, pgl_out and
pgl_fwd) stand before all other iptables rules which ACCEPT traffic.
.IP
To help you achieve this, pglcmd.wd restarts pgl if it detects any problems. But
it's still recommended to restart pgl manually, whenever another application
changed the iptables setup.

.SH TESTING
.PP
See \fIpglcmd test\fR and \fIpglcmd status\fR.
.PP
Or make a traceroute to an IP in the blocklist. If \fBpgl\fR is working all
packets will be dropped at the first hop (your PC), resulting in this output:
.IP
`1 * * *`.
.PP

.SH FILES pgld
.IP /usr/sbin/pgld
The \fBpgl\fR daemon.
.IP /etc/logrotate.d/pgld
Rotates the logfile daily.
.IP /var/log/pgl/pgld.log
The log file of \fBpgld\fR. This file contains amongst other the blocked IPs,
the number of loaded ranges and if the daemon could bind to NFQUEUE.
.IP /var/run/pgld.pid
The pid file of \fBpgld\fR.

.SH FILES and DIRECTORIES pglcmd
.IP /usr/bin/pglcmd
The shell script \fBpglcmd\fR.
.IP /usr/lib/pgl/pglcmd.main
The core of all executable scripts.
.IP /usr/lib/pgl/pglcmd.defaults
Default settings for all variables and pathnames.
.IP /usr/lib/pgl/pglcmd.lib
Library for pglcmd.main.
.IP /etc/pgl/blocklists.list
Lists all blocklists that are used (and downloaded) by \fBpglcmd\fR for use by
pgld.
.IP /etc/pgl/blocklists.local/
Local blocklists are placed here.
.IP /etc/pgl/pglcmd.conf
The user configuration file for \fBpglcmd\fR.
.IP /etc/pgl/allow.p2p
The allow list in the peerguardian .p2p text format.
.IP /etc/NetworkManager/dispatcher.d/20pglcmd
.IP /etc/network/if-up.d/pglcmd
Gets executed whenever a network interface is brought up to make sure that the
automatic whitelisting of LAN traffic and the DNS server does work.
.IP /etc/pgl/[NAME]insert.sh
Script to insert custom iptables rules. Any script in /etc/pgl/ (the
directory is defined in IPTABLES_CUSTOM_DIR) that ends in insert.sh will be
executed on every "pglcmd start", if IPTABLES_SETTINGS="1" or "2" is
configured.
.IP /etc/pgl/[NAME]remove.sh

ChangeLog
0.2 Renaming and documentation

0.1 Initial release
	Versioning the set of scripts used to build and administer my home network
	Working name: ghettobox (from an old thread on the hackmeeting mailinglist)


src/pgld/pgld.c
// pgld - netfilter blocking daemon
//
// Copyright (C) 2009, 2014 Cader <cade.robinson@gmail.com>
// Copyright (C) 2012 hasufell <julian.ospald@googlemail.com>
// Copyright (C) 2011 Dimitris Palyvos-Giannas <jimaras@gmail.com>
// Copyright (C) 2008 Jindrich Makovicka <makovick@gmail.com>
// Copyright (C) 2004 Morpheus <ebutera@users.berlios.de>
//
// This file is part of pgl.
//
// pgl is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// pgl is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with pgl.  If not, see <http://www.gnu.org/licenses/>.

#define PACKAGE_NAME "dowse-pgld"
#define VERSION "1.0"
#define PIDFILE "pgld.pid"
#define CHANNEL "pgl-info-channel"

#include "database.h"
#include "dowse.h"

#include <pgld.h>

static unsigned int accept_mark = 0, reject_mark = 0, use_syslog = 0, queue_length = 0, opt_merge = 0, blockfile_count = 0;
static unsigned short queue_num = 0;
static char *pidfile_name = NULL, *logfile_name=NULL, timestr[17];
static FILE *logfile;
static const char *current_charset = 0, **blocklist_filenames = 0, **blocklist_charsets = 0;
static FILE* pidfile = NULL;

struct nfq_handle *nfqueue_h = 0;
struct nfq_q_handle *nfqueue_qh = 0;

static redisContext *redis = NULL;
static redisReply *reply = NULL;

// General logging function
void do_log(int priority, const char *format, ...) {

    if (use_syslog) {
        va_list ap;
        va_start(ap, format);
        vsyslog(LOG_MAKEPRI(LOG_DAEMON, priority), format, ap);
        va_end(ap);
    }

    if (logfile) {
        va_list ap;
        va_start(ap, format);
        time_t tv;
        struct tm * timeinfo;
        time( &tv );
        timeinfo = localtime ( &tv );
        strftime(timestr, 17, "%b %e %X", timeinfo);
        timestr[16] = '\0';
        fprintf(logfile,"%s ",timestr);
        vfprintf(logfile, format, ap);
        fprintf(logfile, "\n");
        fflush(logfile);
        va_end(ap);
    }

    if (opt_merge) {
        va_list ap;
        va_start(ap, format);
        vfprintf(stderr, format, ap);
        fprintf(stderr,"\n");
        va_end(ap);
    }
}


void int2ip (uint32_t ipint, char *ipstr) {
    ipint=htonl(ipint);
    inet_ntop(AF_INET, &ipint, ipstr, INET_ADDRSTRLEN);
}


static FILE *create_pidfile(const char *name) {
    FILE *f;

    f = fopen(name, "w");
    if (f == NULL) {
        err("Unable to create PID file %s: %s\n", name, strerror(errno));
        return NULL;
    }

    /* this works even if pidfile is stale after daemon is sigkilled */
    if (lockf(fileno(f), F_TLOCK, 0) == -1) {
        err("Unable to set exclusive lock for pidfile %s: %s\n", name, strerror(errno));
        return NULL;
    }

    fprintf(f, "%d\n", getpid());
    fflush(f);

    /* leave fd open as long as daemon is running */
    /* this is useful for example so that inotify can catch a file
     * closed event even if daemon is killed */
    return f;
}

// Once daemonized stdout and stderr are no more available, only logging to
// syslog, logfile.
static void daemonize() {
    /* Fork off and have parent exit. */
    switch (fork()) {
    case -1:
        perror("fork");
        exit(1);
    case 0:
        break;
    default:
        exit(0);
    }

    /* Detach from the controlling terminal */
    setsid();

    /* Close all standard I/0 descriptors */
    close(fileno(stdin));
    close(fileno(stdout));
    close(fileno(stderr));

    act("INFO: Started.");
}

static int load_all_lists() {
    int i, ret = 0;

    blocklist_clear(0);
    if (blockfile_count) {
        for (i = 0; i < blockfile_count; i++) {
            if (load_list(blocklist_filenames[i], blocklist_charsets[i])) {
                err("ERROR: Error loading %s", blocklist_filenames[i]);
                ret = -1;
            }
        }
    } else {
        //assume stdin for list
        load_list(NULL,NULL);
    }
    blocklist_sort();
    blocklist_merge();
    act("INFO: Blocking %u IP ranges (%u IPs).", blocklist.count, blocklist.numips);
    return ret;
}

static void nfqueue_unbind() {
    if (!nfqueue_h)
        return;
    act("INFO: Unbinding from nfqueue: %hu", queue_num);
    nfq_destroy_queue(nfqueue_qh);
    if (nfq_unbind_pf(nfqueue_h, AF_INET) < 0) {
        err("ERROR: Error during nfq_unbind_pf(): %s", strerror(errno));
    }
    nfq_close(nfqueue_h);
}

static void sighandler(int sig, siginfo_t *info, void *context) {
    switch (sig) {
    case SIGUSR1:
        // dump and reset stats
        blocklist_stats(1);
        break;
    case SIGUSR2:
        // just dump stats
        blocklist_stats(0);
        break;
    case SIGHUP:
        if (logfile_name != NULL) {
            act("INFO: Closing logfile: %s", logfile_name);
            fclose(logfile);
            logfile=NULL;
            if ((logfile=fopen(logfile_name,"a")) == NULL) {
                err("ERROR: Unable to open logfile: %s", logfile_name);
                perror(" ");
                exit(-1);
            } else {
                act("INFO: Reopened logfile: %s", logfile_name);
            }
        }
        if (load_all_lists() < 0) {
            err("ERROR: Cannot reload the blocklist(s).");
        }
        act("INFO: Blocklist(s) reloaded.");
        break;
    case SIGTERM:
    case SIGINT:
        nfqueue_unbind();
        blocklist_stats(0);
        blocklist_clear(0);
        free(blocklist_filenames);
        free(blocklist_charsets);
        // close syslog
        if (use_syslog) {
            closelog();
        }
        if (pidfile) {
            fclose(pidfile);
            unlink(pidfile_name);
        }
        exit(0);
        break;
    case SIGSEGV:
        nfqueue_unbind();
        abort();
        break;
    default:
        break;
    }
}

static int install_sighandler() {
    struct sigaction sa;

    memset(&sa, 0, sizeof(sa));
    sa.sa_sigaction = sighandler;
    sa.sa_flags = SA_RESTART | SA_SIGINFO;

    if (sigaction(SIGUSR1, &sa, NULL) < 0) {
        perror("Error setting signal handler for SIGUSR1\n");
        return -1;
    }
    if (sigaction(SIGUSR2, &sa, NULL) < 0) {
        perror("Error setting signal handler for SIGUSR2\n");
        return -1;
    }
    if (sigaction(SIGHUP, &sa, NULL) < 0) {
        perror("Error setting signal handler for SIGHUP\n");
        return -1;
    }
    if (sigaction(SIGTERM, &sa, NULL) < 0) {
        perror("Error setting signal handler for SIGTERM\n");
        return -1;
    }
    if (sigaction(SIGINT, &sa, NULL) < 0) {
        perror("Error setting signal handler for SIGINT\n");
        return -1;
    }
    if (sigaction(SIGSEGV, &sa, NULL) < 0) {
        perror("Error setting signal handler for SIGABRT\n");
        return -1;
    }
    return 0;
}

static void setipinfo (char *src, char *dst, char *proto, struct iphdr *ip, char *payload){
    struct udphdr *udp;
    struct tcphdr *tcp;
    switch (ip->protocol) {
    case TCP:
        strcpy(proto, "TCP");
        tcp     = (struct tcphdr*) (payload + (4 * ip->ihl));
        sprintf(src, "%u.%u.%u.%u:%u",NIPQUAD(ip->saddr),ntohs(tcp->source));
        sprintf(dst, "%u.%u.%u.%u:%u",NIPQUAD(ip->daddr),ntohs(tcp->dest));
        break;
    case UDP:
        strcpy(proto, "UDP");
        udp     = (struct udphdr*) (payload + (4 * ip->ihl));
        sprintf(src, "%u.%u.%u.%u:%u",NIPQUAD(ip->saddr),ntohs(udp->source));
        sprintf(dst, "%u.%u.%u.%u:%u",NIPQUAD(ip->daddr),ntohs(udp->dest));
        break;
    case ICMP:
        strcpy(proto, "ICMP");\
        sprintf(src, "%u.%u.%u.%u",NIPQUAD(ip->saddr));
        sprintf(dst, "%u.%u.%u.%u",NIPQUAD(ip->daddr));
        break;
    default:
        sprintf(proto, "%d", ip->protocol);
        sprintf(src, "%u.%u.%u.%u",NIPQUAD(ip->saddr));
        sprintf(dst, "%u.%u.%u.%u",NIPQUAD(ip->daddr));
        break;
    }
}


static int nfqueue_cb(struct nfq_q_handle *qh, struct nfgenmsg *nfmsg, struct nfq_data *nfa, void *data) {
	int id = 0, status = 0;
	struct nfqnl_msg_packet_hdr *ph;
	block_entry_t *found_range;
	struct iphdr *ip;
	unsigned char *payload;
	char proto[5], src[23], dst[23];  //src and dst are 23 for IP(16)+port(5) + : + NULL
	uint32_t saddr, daddr;
	do_log(LOG_INFO," called nfqueue_cb");
	ph = nfq_get_msg_packet_hdr(nfa);
	if (ph) {
		id = ntohl(ph->packet_id);
		nfq_get_payload(nfa, &payload);
		ip = (struct iphdr*) payload;
		saddr = ntohl(ip->saddr);
		daddr = ntohl(ip->daddr);

		switch (ph->hook) {

		case NF_IP_LOCAL_IN:

			found_range = blocklist_find(saddr);
			if (found_range) {
				struct timeval tv;
				gettimeofday (&tv, NULL);

				// we set the NF_DROP verdict, packet is dropped here
				nfq_set_verdict2(qh, id, NF_DROP, reject_mark, 0, NULL);
				found_range->hits++;
				// NOTE: "setipinfo" sets the formats.
				// TODO: Separate IP and port in there.
				// setipinfo(src, dst, proto, ip, payload);
				
				cmd_redis(redis, "publish %s PGL,%u.%u.%u.%u,BLOCK,%lu,%u.%u.%u.%u",
				        CHANNEL, NIPQUAD(ip->daddr), tv.tv_sec, NIPQUAD(ip->saddr) );

				// do_log(LOG_NOTICE, " IN: %-22s %-22s %-4s || %s",src,dst,proto,found_range->name);


			// } else {
			// 	// we set the user-defined accept_mark and set NF_REPEAT verdict
			// 	// it's up to other iptables rules to decide what to do with this marked packet
				// nfq_set_verdict2(qh, id, NF_REPEAT, accept_mark, 0, NULL);
			}
			break;

		case NF_IP_LOCAL_OUT:

			found_range = blocklist_find(daddr);
			if (found_range) {
				struct timeval tv;
				gettimeofday (&tv, NULL);

				// we set the user-defined reject_mark and set NF_REPEAT verdict
				// it's up to other iptables rules to decide what to do with this marked packet
				nfq_set_verdict2(qh, id, NF_DROP, reject_mark, 0, NULL);
				found_range->hits++;
				// NOTE: "setipinfo" sets the formats.
				// TODO: Separate IP and port in there.
				// setipinfo(src, dst, proto, ip, payload);
				cmd_redis(redis, "publish %s PGL,%u.%u.%u.%u,BLOCK,%lu,%u.%u.%u.%u",
				        CHANNEL, NIPQUAD(ip->saddr), tv.tv_sec, NIPQUAD(ip->daddr) );
				// do_log(LOG_NOTICE, "OUT: %-22s %-22s %-4s || %s",src,dst,proto,found_range->name);

			// } else {
			// 	// we set the user-defined accept_mark and set NF_REPEAT verdict
			// 	// it's up to other iptables rules to decide what to do with this marked packet
			// 	nfq_set_verdict2(qh, id, NF_REPEAT, accept_mark, 0, NULL);
			}
			break;
		case NF_IP_FORWARD:
		  do_log(LOG_INFO,"Looking for %x",daddr);
			found_range = blocklist_find(daddr);
			if (!found_range) {
				found_range = blocklist_find(saddr);
			}
			if (found_range) {
				struct timeval tv;
				gettimeofday (&tv, NULL);

				// we set the user-defined reject_mark and set NF_REPEAT verdict
				// it's up to other iptables rules to decide what to do with this marked packet
				nfq_set_verdict2(qh, id, NF_DROP, reject_mark, 0, NULL);
				found_range->hits++;
				// setipinfo(src, dst, proto, ip, payload);

				cmd_redis(redis, "publish %s PGL,%u.%u.%u.%u,BLOCK,%lu,%u.%u.%u.%u",
				        CHANNEL, NIPQUAD(ip->saddr), tv.tv_sec, NIPQUAD(ip->daddr) );
				// do_log(LOG_NOTICE, "FWD: %-22s %-22s %-4s || %s",src,dst,proto,found_range->name);

			// } else {
			// 	// we set the user-defined accept_mark and set NF_REPEAT verdict
			// 	// it's up to other iptables rules to decide what to do with this marked packet
			// 	nfq_set_verdict2(qh, id, NF_REPEAT, accept_mark, 0, NULL);
			}
			break;
		default:
			 do_log(LOG_NOTICE, "WARN: Not NF_LOCAL_IN/OUT/FORWARD packet!");
			break;
		}
	} else {
		do_log(LOG_ERR, "ERROR: NFQUEUE: can't get msg packet header.");
		return 1;               // from nfqueue source: 0 = ok, >0 = soft error, <0 hard error
	}
	return 0;
}

static int nfqueue_bind() {
    nfqueue_h = nfq_open();
    if (!nfqueue_h) {
        do_log(LOG_ERR, "ERROR: Error during nfq_open(): %s", strerror(errno));
        return -1;
    }

    if (nfq_unbind_pf(nfqueue_h, AF_INET) < 0) {
        do_log(LOG_ERR, "ERROR: Error during nfq_unbind_pf(): %s", strerror(errno));
        nfq_close(nfqueue_h);
        return -1;
    }

    if (nfq_bind_pf(nfqueue_h, AF_INET) < 0) {
        do_log(LOG_ERR, "ERROR: Error during nfq_bind_pf(): %s", strerror(errno));
        nfq_close(nfqueue_h);
        return -1;
    }

    do_log(LOG_INFO, "INFO: Binding to queue %hu", queue_num);
    if (accept_mark) {
        do_log(LOG_INFO, "INFO: ACCEPT mark: %u", accept_mark);
    }
    if (reject_mark) {
        do_log(LOG_INFO, "INFO: REJECT mark: %u", reject_mark);
    }
    nfqueue_qh = nfq_create_queue(nfqueue_h, queue_num, &nfqueue_cb, NULL);
    if (!nfqueue_qh) {
        do_log(LOG_ERR, "ERROR: Error during nfq_create_queue(): %s", strerror(errno));
        nfq_close(nfqueue_h);
        return -1;
    }

    if (nfq_set_mode(nfqueue_qh, NFQNL_COPY_PACKET, PAYLOADSIZE) < 0) {
        do_log(LOG_ERR, "ERROR: Can't set packet_copy mode: %s", strerror(errno));
        nfq_destroy_queue(nfqueue_qh);
        nfq_close(nfqueue_h);
        return -1;
    }

    if ( queue_length > 0) {
        if (nfq_set_queue_maxlen(nfqueue_qh, queue_length) < 0) {
            do_log(LOG_ERR, "ERROR: Can't set queue max length: %s", strerror(errno));
            do_log(LOG_INFO, "INFO: Continuing anyway with default queue max length.");
        } else {
            do_log(LOG_INFO, "INFO: Set netfilter queue length to %u packets", queue_length);
        }
    }
    return 0;
}

static void nfqueue_loop () {
    struct nfnl_handle *nh;
    int fd, rv;
    char buf[RECVBUFFSIZE];
//  struct pollfd fds[1];

    do_log(LOG_INFO, " into %s %d ",__FUNCTION__,__LINE__);
    if (nfqueue_bind() < 0) {
        do_log(LOG_ERR, "ERROR: Error binding to queue: %hu", queue_num);
        exit(1);
    }

    nh = nfq_nfnlh(nfqueue_h);
    do_log(LOG_INFO, "  %s %d ",__FUNCTION__,__LINE__);
    
    fd = nfnl_fd(nh);
    do_log(LOG_INFO, "  %s %d ",__FUNCTION__,__LINE__);
    
    while ((rv = recv(fd, buf, sizeof(buf), 0)) >= 0) {
      do_log(LOG_INFO, " into while loop %s %d ",__FUNCTION__,__LINE__);
        nfq_handle_packet(nfqueue_h, buf, rv);
    }
    int err=errno;
    if ( err == ENOBUFS ) {
        do_log(LOG_ERR, "ERROR: ENOBUFS error on queue '%hu'. Use pgld -Q option or set in pglcmd.conf NFQUEUE_MAXLEN to increase buffers, recv returned %s", queue_num, strerror(err));
        /* close and return, nfq_destroy_queue() won't work as we've no buffers */
        nfq_close(nfqueue_h);
        exit(1);
    } else {
        do_log(LOG_ERR, "ERROR: Error on queue '%hu', recv returned %s", queue_num, strerror(err));
        nfqueue_unbind();
        exit(0);
    }
}

static void print_usage() {
    fprintf(stderr, PACKAGE_NAME " " VERSION "\n\n");
    fprintf(stderr, "pgl is licensed under the GNU General Public License v3, or (at your option)\n");
    fprintf(stderr, "any later version. This program comes with ABSOLUTELY NO WARRANTY. This is\n");
    fprintf(stderr, "free software, and you are welcome to modify and/or redistribute it.\n\n");
    fprintf(stderr, "Usage:\n");
    fprintf(stderr, "  pgld [-s] [-l LOGFILE]");

    fprintf(stderr, " [-c CHARSET] [-p PIDFILE] -a MARK -r MARK [-q 0-65535] [-Q queue_size] BLOCKLIST(S)\n");
    fprintf(stderr, "  pgld [-c CHARSET] -m [BLOCKLIST(S)]\n\n");
    fprintf(stderr, "Options:\n");
    fprintf(stderr, "  -s                Enable logging to syslog.\n");
    fprintf(stderr, "  -l LOGFILE        Enable logging to LOGFILE.\n");

    fprintf(stderr, "  -c CHARSET        Specify blocklist file CHARSET.\n");
    fprintf(stderr, "  -p PIDFILE        Use a PIDFILE.\n");
    fprintf(stderr, "  -q 0-65535        Specify a 16-bit NFQUEUE number.\n");
    fprintf(stderr, "                    Must match --queue-num used in iptables rules.\n");
    fprintf(stderr, "  -Q queue_size     Specify a queue length in packets. If not specified use default.\n");
    fprintf(stderr, "  -r MARK           Place a 32-bit MARK on rejected packets.\n");
    fprintf(stderr, "  -a MARK           Place a 32-bit MARK on accepted packets.\n");
    fprintf(stderr, "  -m [BLOCKLIST(S)] Load, sort, merge, and dump blocklist(s) to stdout.\n");
    fprintf(stderr, "                    Specify blocklists or pipe from stdin.\n");
    fprintf(stderr, "  -h                Print this help.\n");
}

void add_blocklist(const char *name, const char *charset) {
    blocklist_filenames = (const char**)realloc(blocklist_filenames, sizeof(const char*) * (blockfile_count + 1));
    CHECK_OOM(blocklist_filenames);
    blocklist_charsets = (const char**)realloc(blocklist_charsets, sizeof(const char*) * (blockfile_count + 1));
    CHECK_OOM(blocklist_charsets);
    blocklist_filenames[blockfile_count] = name;
    blocklist_charsets[blockfile_count] = charset;
    blockfile_count++;
}

int main(int argc, char *argv[]) {
    int opt, i;
    while ((opt = getopt(argc, argv, "sl:c:p:q:Q:r:a:mh" )) != -1) {
        switch (opt) {
        case 's':
            use_syslog = 1;
            break;
        case 'l':
            logfile_name=malloc(strlen(optarg)+1);
            CHECK_OOM(logfile_name);
            strcpy(logfile_name,optarg);
            break;
        case 'c':
            current_charset = optarg;
            break;
        case 'p':
            pidfile_name=malloc(strlen(optarg)+1);
            CHECK_OOM(pidfile_name);
            strcpy(pidfile_name,optarg);
            break;
        case 'q':
            queue_num = (uint16_t)atoi(optarg);
            break;
        case 'Q':
            queue_length = (uint32_t)atoi(optarg);
            break;
        case 'r':
            reject_mark = (uint32_t)atoi(optarg);
            break;
        case 'a':
            accept_mark = (uint32_t)atoi(optarg);
            break;
        case 'm':
            opt_merge = 1;
            break;
        case 'h':
            print_usage();
            exit(0);
        }
    }

    for (i = 0; i < argc - optind; i++) {
        add_blocklist(argv[optind + i], current_charset);
    }

    // Do merge and exit - for just merging lists and not daemonizing.
    if (opt_merge) {
        blocklist_init();
        load_all_lists();
        blocklist_dump();
        exit(0);
    }

    if (blockfile_count == 0) {
        fprintf(stderr, "\nERROR: No blocklist specified!\n\n");
        print_usage();
        exit(1);
    }

    if (queue_num < 0 || queue_num > 65535) {
        fprintf(stderr, "\nERROR: Invalid queue number! Must be 0-65535\n\n");
        print_usage();
        exit(1);
    }

    if (pidfile_name == NULL) {
        pidfile_name=malloc(strlen(PIDFILE)+1);
        CHECK_OOM(pidfile_name);
        strcpy(pidfile_name,PIDFILE);
    }

    // open redis connection
    redis = connect_redis(db_dynamic);

    // open logfile
    if (logfile_name != NULL) {
        if ((logfile=fopen(logfile_name,"a")) == NULL) {
            fprintf(stderr, "Unable to open logfile: %s", logfile_name);
            print_usage();
            perror(" ");
            exit(-1);
        }
    }

    // open syslog
    if (use_syslog) {
        openlog("pgld", 0, LOG_DAEMON);
    }

    daemonize();

    // pidfile has to be created *after* daemonize (results in a new pid)
    pidfile = create_pidfile(pidfile_name);
    if (!pidfile) {
        do_log(LOG_ERR, "ERROR: Error creating pidfile %s", pidfile_name);
        exit(1);
    }

    if (install_sighandler() != 0) {
        do_log(LOG_ERR, "ERROR: Error installing signal handlers.");
        exit(1);
    }

    blocklist_init();
    if (load_all_lists() < 0) {
        do_log(LOG_ERR, "ERROR: Cannot load the blocklist(s)");
        return -1;
    }

    nfqueue_loop();
    return(0);
}
